pkg/services/alertmanager.go

```diff
@@ -0,0 +1,207 @@
+		opts.APIPath = "/api/v2/alerts"
+	}
+
+	return &alertmanagerService{entry: log.WithField("service", "alertmanager"), opts: opts}
+}
+
+type alertmanagerService struct {
+	entry *log.Entry
+	opts  AlertmanagerOptions
+}
+
+// GetTemplater parse text template
+func (n AlertmanagerNotification) GetTemplater(name string, f texttemplate.FuncMap) (Templater, error) {
+	return func(notification *Notification, vars map[string]interface{}) error {
+		if notification.Alertmanager == nil {
+			notification.Alertmanager = &AlertmanagerNotification{}
+		}
+		notification.Alertmanager.StartsAt = time.Now()
+
+		tmplGeneratorURL := n.GeneratorURL
+		if tmplGeneratorURL == "" {
+			tmplGeneratorURL = "{{.app.spec.source.repoURL}}"
+		}
+		tmpl, err := texttemplate.New(name).Funcs(f).Parse(tmplGeneratorURL)
+		if err != nil {
+			return err
+		}
+		var tempData bytes.Buffer
+		if err := tmpl.Execute(&tempData, vars); err != nil {
+			return err
+		}
+		if val := tempData.String(); val != "" {
+			notification.Alertmanager.GeneratorURL = val
+		}
+		// generatorURL must url format
+		if _, err := url.Parse(notification.Alertmanager.GeneratorURL); err != nil {
+			return err
+		}
+
+		if len(n.Labels) > 0 {
+			notification.Alertmanager.Labels = n.Labels
+			if err := notification.Alertmanager.parseLabels(name, f, vars); err != nil {
+				return err
+			}
+		}
+		if len(n.Annotations) > 0 {
+			notification.Alertmanager.Annotations = n.Annotations
+			if err := notification.Alertmanager.parseAnnotations(name, f, vars); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}, nil
+}
+
+func (n *AlertmanagerNotification) parseAnnotations(name string, f texttemplate.FuncMap, vars map[string]interface{}) error {
+	for k, v := range n.Annotations {
+		var tempData bytes.Buffer
+		tmpl, err := texttemplate.New(name).Funcs(f).Parse(v)
+		if err != nil {
+			return err
+		}
+		if err := tmpl.Execute(&tempData, vars); err != nil {
+			return err
+		}
+		if val := tempData.String(); val != "" {
+			n.Annotations[k] = val
+		}
+	}
+
+	return nil
+}
+
+func (n *AlertmanagerNotification) parseLabels(name string, f texttemplate.FuncMap, vars map[string]interface{}) error {
+	foundAlertname := false
+
+	for k, v := range n.Labels {
+		if k == alertNameLabel {
+			foundAlertname = true
+		}
+
+		var tempData bytes.Buffer
+		tmpl, err := texttemplate.New(name).Funcs(f).Parse(v)
+		if err != nil {
+			return err
+		}
+		if err := tmpl.Execute(&tempData, vars); err != nil {
+			return err
+		}
+		if val := tempData.String(); val != "" {
+			n.Labels[k] = val
+		}
+	}
+
+	if !foundAlertname {
+		n.Labels[alertNameLabel] = name
+	}
+
+	return nil
+}
+
+// Send using create alertmanager events
+func (s alertmanagerService) Send(notification Notification, dest Destination) error {
+	rawBody, err := json.Marshal([]*AlertmanagerNotification{notification.Alertmanager})
+	if err != nil {
+		return err
+	}
+
+	for _, target := range s.opts.Targets {
+		go func(target string) {
```

Comment:
Would you check the end of goroutine using sync.WaitGroup or something?

Reply:
The alertmanager high availability cluster provides multiple instance addresses externally. The client does not need to ensure that each instance api called is successfully received. As long as one instance is successfully received, the internal sync status will be between them.